README.md
# Kadira

Handle core functionalities of Kadira such as

* Transport
* NTP Time syncing

### Kadira Transport Debugging

You can debug what's happening inside the kadira transport by exposing following environment variable:

```
export `DEBUG=kadira-core:transport`
```


CHANGELOG.md
# Change Log

### v1.3.2

* Remove isomorphic-fetch and implements our own isomorphism.

### v1.3.1
* Handle HTTP request errors. These are the errors coming from the request module.

### v1.3.0

* Add getJob() api

### v1.2.0

* Remove bluebird and just use babel's Promise

### v1.1.0

* Using request instead of isomorphic-fetch as the transport

### v1.0.0

* First major release


lib/__tests__/clock.js
import assert from 'assert';
import {describe, it, beforeEach, afterEach} from 'mocha';
import server from './_server.js';
import Clock from '../clock.js';

describe('clock', function () {
  const endpoint = 'http://localhost:8000/simplentp/sync';

  function inRange(n, min, max) {
    return n <= max && n >= min;
  }

  beforeEach(function (done) {
    server.setCount(0);
    return server.start(done);
  });

  afterEach(function (done) {
    return server.stop(done);
  });

  it('should set diff value', async function () {
    const clock = new Clock({endpoint});
    await clock.sync();
    assert(inRange(clock._diff, -1100, -900));
  });

  it('should return fixed current time', async function () {
    const clock = new Clock({endpoint});
    await clock.sync();
    assert(inRange(clock.getTime() - Date.now(), -1100, -900));
  });

  it('should return fixed timestamps', async function () {
    const clock = new Clock({endpoint});
    await clock.sync();
    assert(inRange(clock.fixTime(1e4) - 1e4, -1100, -900));
  });
});


lib/__tests__/index.js
import assert from 'assert';
import {describe, it, beforeEach, afterEach} from 'mocha';
import server from './_server.js';
import Kadira from '../index.js';
import {ByPassRetryError} from '../retry.js';

describe('kadira', function () {
  const endpoint = 'http://localhost:8000';
  const validAuth = {appId: 'test-app-id', appSecret: 'test-app-secret'};
  const validOpts = Object.assign({endpoint}, validAuth);
  const invldOpts = Object.assign({endpoint});

  function inRange(n, min, max) {
    return n <= max && n >= min;
  }

  beforeEach(function (done) {
    server.setCount(0);
    return server.start(done);
  });

  afterEach(function (done) {
    return server.stop(done);
  });

  describe('connect', function () {
    it('should throw with wrong info', async function () {
      let erred = false;
      const kadira = new Kadira(invldOpts);

      try {
        await kadira.connect();
      } catch (e) {
        assert.equal(e instanceof ByPassRetryError, true);
        erred = true;
      }

      assert(erred);
      kadira.disconnect();
    });

    it('should connect with correct info', async function () {
      const kadira = new Kadira(validOpts);
      await kadira.connect();
      kadira.disconnect();
    });

    it('should sync the diff value', async function () {
      const kadira = new Kadira(validOpts);
      await kadira.connect();
      kadira.disconnect();
      assert(inRange(kadira._clock._diff, -1100, -900));
    });
  });

  describe('sendData', function () {
    it('should send data to the server', async function () {
      const options = Object.assign({}, validOpts, {dataFlushInterval: 100});
      const kadira = new Kadira(options);
      await kadira.connect();
      kadira.disconnect();

      await kadira.sendData({
        test1: [ {a: 'b'}, {c: 'd'} ],
        test2: [ {e: 'f'} ],
      });

      assert.deepEqual(server.getData(), {
        host: kadira._options.hostname,
        test1: [ {a: 'b'}, {c: 'd'} ],
        test2: [ {e: 'f'} ],
      });
    });
  });

  describe('updateJob', function () {
    it('should send data to the server', async function () {
      const kadira = new Kadira(validOpts);
      await kadira.connect();
      kadira.disconnect();

      await kadira.updateJob('job-0', {foo: 'bar'});
      assert.deepEqual(server.getJobs(), {
        action: 'set',
        params: {id: 'job-0', foo: 'bar'},
      });
    });
  });

  describe('getJob', function () {
    it('should send data to the server', async function () {
      const kadira = new Kadira(validOpts);
      await kadira.connect();
      kadira.disconnect();

      const res = await kadira.getJob('job-0');
      assert.deepEqual(res, {aa: 10});
      assert.deepEqual(server.getJobs(), {
        action: 'get',
        params: {id: 'job-0'},
      });
    });
  });

  describe('_checkAuth', () => {
    describe('with correct login info', () => {
      it('should just return', done => {
        const kadira = new Kadira(validOpts);
        kadira._checkAuth().then(() => {
          done();
        });
      });
    });

    describe('with bad login info', () => {
      it('should throw an error', done => {
        const kadira = new Kadira(invldOpts);
        kadira._checkAuth().catch(err => {
          assert.equal(err.message, 'Unauthorized');
          done();
        });
      });
    });
  });
});


lib/__tests__/retry.js
import assert from 'assert';
import {describe, it} from 'mocha';
import retry from '../retry.js';
import {ByPassRetryError, MaxRetryError} from '../retry.js';

describe('retry', function () {
  it('should retry N times', async function () {
    this.timeout(6e4);
    let erred = false;
    let result;
    const err = new Error('Some Error');

    try {
      result = await retry(() => {
        return new Promise((_, r) => r(err));
      });
    } catch (e) {
      assert.equal(e instanceof MaxRetryError, true);
      assert.equal(e.message, `Reached maximum retry limit for ${err.message}`);
      erred = true;
    }

    assert(erred);
    assert.equal(result, undefined);
  });

  it('should stop on success', async function () {
    this.timeout(6e4);
    let erred = false;
    let result;

    try {
      result = await retry(() => Promise.resolve('result'));
    } catch (e) {
      erred = true;
    }

    assert(!erred);
    assert.equal(result, 'result');
  });

  it('should stop on endretry', async function () {
    this.timeout(6e4);
    let erred = false;
    let result;

    try {
      result = await retry(() => {
        return new Promise((_, reject) => reject(new ByPassRetryError('ERR')));
      });
    } catch (e) {
      assert.equal(e instanceof ByPassRetryError, true);
      assert.equal(e.message, 'ERR');
      erred = true;
    }

    assert(erred);
    assert.equal(result, undefined);
  });
});


lib/clock.js
/* global fetch */

import retry from './retry.js';
let fetchUrl;
if (typeof fetch === 'function') {
  // has a native fetch implementation
  fetchUrl = fetch;
} else if (typeof __dirname !== 'undefined') {
  // for Node.js
  fetchUrl = require('node-fetch');
  fetchUrl.Promise = Promise;
} else {
  // for the browser
  fetchUrl = require('whatwg-fetch');
}

const DEFAULTS = {
  endpoint: '',
};

export default class Clock {
  constructor(_options) {
    this._options = Object.assign({}, DEFAULTS, _options);
    this._diff = 0;
    this.ready = false;
  }

  getTime() {
    // current time on Kadira server
    return this._clientTS() + this._diff;
  }

  fixTime(timestamp) {
    // `timestamp` on Kadira server
    return timestamp + this._diff;
  }

  sync() {
    // calculate the time difference
    return retry(() => this._syncOnce());
  }

  _clientTS() {
    // Get client timestamp while considering time related
    // libraries messing up the Date object. "Y U DO THIS?"
    let now = Date.now();
    if (typeof now === 'number') {
      return now;
    }

    // some time related libraries screw up Date.now()
    // and it returns a Date object instead of a number
    if (now instanceof Date) {
      return now.getTime();
    }

    // final attempt to get time
    return new Date().getTime();
  }

  _syncOnce() {
    let startTS;

    return this._fetchTime().then(() => {
      startTS = this._clientTS();
      return this._fetchTime();
    }).then(serverTS => {
      const latency = (this._clientTS() - startTS) / 2;
      this._diff = serverTS - latency - startTS;
      this.ready = true;
    });
  }

  _fetchTime() {
    return fetchUrl(this._options.endpoint).then(res => {
      if (res.status !== 200) {
        throw new Error('request failed: ' + res.status);
      }

      return res.text().then(txt => parseInt(txt, 10));
    });
  }
}


lib/retry.js
import {inherits} from 'util';

// default options
const options = {
  maxRetries: 4,
  timeFunction: i => 100 * Math.pow(i, 2),
};

// XXX: We we need to use instanceof with these error classes.
// So, we can't use Babel's version of extends for that.
// It's doesn't support that. See: http://stackoverflow.com/a/33877501/457224
// That's why we are doing it in the old fashion way.

// reject the promise with this error when run out of retry attmpts.
export const MaxRetryError = function (message) {
  Error.call(this, message);
  this.message = message;
};
inherits(MaxRetryError, Error);

// reject the promise with this error (in promiser) to stop retrying.
export const ByPassRetryError = function (message) {
  Error.call(this, message);
  this.message = message;
};
inherits(MaxRetryError, Error);

// retry([options], fn)
// retry module takes a `promiser` function as the main argument.
// The promiser function should return a promise which will be used
// to decide whether the task ran successfully. If the task failed
// it will retry by running the `promiser` function again. Retry will
// stop when it has tried `maxRetries` times or if the promise fails
// with the special error `ERR_ENDRETRY`.
export default function retry(promiser) {
  // The retry module returns a promise which will end when the task
  // is successful or when the retry fails by retry count or by user.
  // It will also collect start/end times for each retry attempt.
  return new Promise(function (resolve, reject) {
    let count = 0;

    const onError = function (err) {
      if (err instanceof ByPassRetryError) {
        reject(err);
      } else {
        attempt(err);
      }
    };

    const attempt = async function (lastError) {
      if (++count > options.maxRetries) {
        const message = `Reached maximum retry limit for ${lastError.message}`;
        const err = new MaxRetryError(message);
        return reject(err);
      }

      // stop a few milliseconds between retries
      const millis = options.timeFunction(count);
      await delay(millis);

      promiser()
        .then(resolve, onError);
    };

    // start!
    attempt();
  });
}

function delay(millis) {
  return new Promise(resolve => {
    setTimeout(resolve, millis);
  });
}


src/courier.js
import fetch from 'isomorphic-fetch';

// return error after retries
export const MAX_RETRIES = 5;

export default class Courier {
  constructor(uri) {
    this._uri = uri;
  }

  send(data) {
    return new Promise((resolve, reject) => {
      let retries = 0;
      let body;

      try {
        body = JSON.stringify(data);
      } catch (e) {
        reject(e);
      }

      // try to send data to the server
      // check retry limit before trying
      const attempt = () => {
        if (retries++ >= MAX_RETRIES) {
          reject(new Error('reached maximum retry limit'));
          return;
        }

        // backoff 0, 500, 2000, 4500, 8000, ...
        const millis = 500 * Math.pow(retries, 2);

        setTimeout(() => {
          fetch(this._uri, {method: 'POST', body})
            .then(validate)
            .catch(attempt);
        }, millis);
      };

      // validate response from the server
      // if necessary, try sending again
      const validate = res => {
        if (res.status === 200) {
          res.json().then(resolve);
        } else if (res.status >= 400 && res.status < 500) {
          reject(new Error('failed by client error: ' + res.status));
        } else {
          attempt();
        }
      };

      attempt();
    });
  }
}


lib/index.js
import request from 'request';
import Clock from './clock.js';
import retry from './retry.js';
import {ByPassRetryError} from './retry.js';
import debug from 'debug';

const logger = debug('kadira-core:transport');

const DEFAULTS = {
  appId: '',
  appSecret: '',
  endpoint: 'https://enginex.kadira.io',
  hostname: 'localhost',
  clockSyncInterval: 1000 * 60,
  dataFlushInterval: 1000 * 10,
};

// exporting this for if we need to get this as a NPM module.
export class Kadira {
  constructor(_options) {
    this._options = Object.assign({}, DEFAULTS, _options);
    this._headers = {
      'content-type': 'application/json',
      accepts: 'application/json',
      'KADIRA-APP-ID': this._options.appId,
      'KADIRA-APP-SECRET': this._options.appSecret,
    };

    this._clock = new Clock({
      endpoint: this._options.endpoint + '/simplentp/sync',
    });

    this._clockSyncInterval = null;
  }

  connect() {
    logger('connecting with', this._options);
    return this._checkAuth()
      .then(() => this._clock.sync())
      .then(() => {
        this._clockSyncInterval = setInterval(
          () => this._clock.sync(),
          this._options.clockSyncInterval
        );
      });
  }

  disconnect() {
    logger('disconnect');
    clearInterval(this._clockSyncInterval);
  }

  getJob(id) {
    const body = {action: 'get', params: {}};
    Object.assign(body.params, {id});

    const url = this._options.endpoint + '/jobs';
    const params = {
      body,
      headers: this._headers,
      json: true
    };

    logger('get job', id);
    return this._send(url, params);
  }

  updateJob(id, diff) {
    const body = {action: 'set', params: {}};
    Object.assign(body.params, diff, {id});

    const url = this._options.endpoint + '/jobs';
    const params = {
      body,
      headers: this._headers,
      json: true
    };

    logger('update job', id);
    return this._send(url, params);
  }

  // send the given payload to the server
  sendData(_payload) {
    const payload = {
      ..._payload,
      host: this._options.hostname
    };

    const url = this._options.endpoint;
    const body = JSON.stringify(payload);
    const params = {
      body,
      headers: this._headers,
    };

    logger(`send data - ${body.substr(0, 50)}...`);
    return this._send(url, params);
  }

  // ping the server to check whether appId and appSecret
  // are valid and correct. Data sent inside http headers.
  _checkAuth() {
    const uri = this._options.endpoint + '/ping';
    const params = {headers: this._headers};
    return this._send(uri, params);
  }

  // communicates with the server with http (using fetch)
  // Also handles response http status codes and  retries
  _send(url, params) {
    return retry(() => {
      return new Promise((resolve, reject) => {
        request.post(url, params, (err, res, body) => {
          if (err) {
            return reject(err);
          }

          if (res.statusCode === 200) {
            return resolve(body);
          }

          if (res.statusCode === 401) {
            logger('Error: Unauthorized');
            return reject(new ByPassRetryError('Unauthorized'));
          }

          if (res.statusCode >= 400 && res.statusCode < 500) {
            const message = `Agent Error: ${res.statusCode}`;
            logger(`Error: ${message}`);
            return reject(new ByPassRetryError(message));
          }

          const message = `Request failed: ${res.statusCode}`;
          logger(`Error: ${message}`);
          reject(new Error(message));
        });
      });
    });
  }
}

export default Kadira;
